File: day24/src/main/cpp/solution.cpp
#include <algorithm>
#include <fstream>
#include <iostream>
#include <list>
#include <map>
#include <string>
#include <vector>

using namespace std;

struct variables {
  long long xAdd;
  long long zDiv;
  long long yAdd;
};

vector<variables> input;
vector<long long> zValues;
map<pair<int, long long>, vector<list<int>>> cache;

void readInputFile() {
  ifstream file("../resources/input");
  string s;
  int count = 0;
  long long xval, yval, zval;
  while (getline(file, s)) {
    if (count % 18 == 4) {
      zval = stoll(s.substr(6, s.length()));
    } else if (count % 18 == 5) {
      xval = stoll(s.substr(6, s.length()));
    } else if (count % 18 == 15) {
      yval = stoll(s.substr(6, s.length()));
      input.emplace_back(variables{xval, zval, yval});
    }
    ++count;
  }
  for (int i = 13; i >= 0; --i) {
    auto var = input[i];
    zval = var.zDiv;
    if (zValues.size() == 0) {
      zValues.emplace_back(zval);
    } else {
      zValues.emplace_back(zval * zValues.back());
    }
  }
  reverse(zValues.begin(), zValues.end());
}

long long checkDigit(long long w, long long zVal, variables vars) {
  long long x = vars.xAdd + (zVal % 26);
  long long z = zVal / vars.zDiv;
  if (x != w) {
    z = z * 26 + w + vars.yAdd;
  }
  return z;
}

vector<list<int>> smartSearch(int index, long long z) {
  auto key = pair(index, z);
  if (cache.find(pair(index, z)) != cache.end()) {
    return cache[key];
  }
  vector<list<int>> result;
  if (index == 14) {
    if (z == 0) {
      result.push_back({});
    }
  } else if (z < zValues[index]) {
    auto vars = input[index];
    for (int w = 1; w < 10; ++w) {
      auto next = checkDigit(w, z, vars);
      for (auto option : smartSearch(index + 1, next)) {
        option.push_front(w);
        result.push_back(option);
      }
    }
  }

  cache.insert(pair(key, result));
//   if (result.size() > 0) {
//     cout << key.first << "," << key.second << ": " << result.size() << endl;
//   }
  return result;
}

int main() {
  readInputFile();
  auto options = smartSearch(0, 0);
  string s;
  long long minValue  = INT64_MAX;
  long long maxValue = INT32_MIN;
  for(auto const &option: options) {
      if(option.size() == 14) {
          s = "";
          for(auto const digit: option) {
              s += '0'+ digit;
          }
          minValue = min(stoll(s), minValue);
          maxValue = max(stoll(s), maxValue);
      }
  }
    cout << "part 1" << endl
         << maxValue << endl
         << "part 2" << endl
         << minValue << endl;
  return 0;
}

File: day24/src/main/python/solution.py
from functools import cache


def read_input():
    f = open('../resources/input', 'r')
    i = 0
    xVal = yVal = zVal = 0
    for line in f.readlines():
        if i % 18 == 4:
            zVal = int(line[6:-1])
        elif i % 18 == 5:
            xVal = int(line[6:-1])
        elif i % 18 == 15:
            yVal = int(line[6:-1])
            variables.append([zVal, xVal, yVal])
        i += 1
    calculate_z_values()


def calculate_z_values():
    for i in range(0, 14):
        index = 13 - i
        if len(z_values) == 0:
            z_values.append(variables[index][0])
        else:
            z_values.append(variables[index][0] * z_values[-1])
    z_values.reverse()


def check_digit(w, z, xAdd, zDiv, yAdd):
    x = xAdd + z % 26
    z = z // zDiv
    if x != w:
        z = z * 26 + w + yAdd
    return z


@cache
def smart_search(index, z):
    if index == 14:
        if z == 0:
            return [""]
        else:
            return []
    if z > z_values[index]:
        return []
    zDiv, xAdd, yAdd = variables[index]
    result = []
    for w in range(1, 10):
        next_z = check_digit(w, z, xAdd, zDiv, yAdd)
        for option in smart_search(index + 1, next_z):
            result.append(str(w) + option)
    return result


z_values = []
variables = []

if __name__ == "__main__":
    read_input()
    results = smart_search(0, 0)
    print("part 1")
    print(max(results))
    print("part 2")
    print(min(results))
